src/util/treadmill.rs

```diff
@@ -6,111 +6,192 @@ use crate::util::ObjectReference;
 
 use super::object_enum::ObjectEnumerator;
 
+/// A data structure for recording objects in the LOS.
+///
+/// It is divided into the nursery and the mature space, and each of them is further divided into
+/// the from-space and the to-space.
+///
+/// All operations are protected by a single mutex [`TreadMill::sync`].
 pub struct TreadMill {
-    from_space: Mutex<HashSet<ObjectReference>>,
-    to_space: Mutex<HashSet<ObjectReference>>,
-    collect_nursery: Mutex<HashSet<ObjectReference>>,
-    alloc_nursery: Mutex<HashSet<ObjectReference>>,
+    sync: Mutex<TreadMillSync>,
```

Comment:
This `Mutex` can be pushed to the user of `TreadMill`. They can define `Mutex<Treadmill>`.

Reply:
The JikesRVM implementation is not lock free. It is a bit strange to have a struct that its only field is a `Mutex` and every method to the struct will acquire the lock. But it is fine.